Feature/spin and metal support

bosonie opened the pull request:
Introduces the second set of decision taken in the meeting of the 4th of Nov:
* Three new arguments of `get_builder`: `is_insulator`, `spin`, `initial_magnetization` with appropriate options and defaults
* Added the CLI option for spin (both eos and relax command)

Still open question:
* How to implement the common interface to pass the `initial_magnetization`? Should we agree on @zhubonan proposal? Do we create a specific object for the purpose or we just agree on a dictionary with specific keys?

Also a new question came in my mind:
* Shouldn't we now at least return in output the total spin polarization (or magnetization) of the system? Even better would be the final atomic spins.

Also, I decided to centralize the checks on the arguments passed to `get_builder`. This means that the `get_builder` method of the `RelaxInputsGenerator` abstract class already checks that the arguments `relaxation_type`, `is_insulator` and `spin` received an acceptable value. Every child class of  `RelaxInputsGenerator` calls the super of `get_builder`. Few points:
* I could do the same for the `protocol` argument. However here we should decide what to do if a not accepted value is passed. In my siesta implementation I just throw a warning and I use the default protocol. However someone could prefer to raise error.
* We could introduce here checks on `calc_engines`, meaning that it contains `code` and `options` keys. 

Please give your opinion asap, so we can conclude this PR and leave enough time for the plugin specific modifications.

zhubonan commented:
For the magnetisation initialisation, @bosonie maybe we should built-in some functions/class to allow convert between the two ways of representing the moments - by dictionary mapping based on kinds as well as the raw spins per atoms?

bosonie commented:
@sphuber, can you help me to understand what happens for `aiida_quantumespresso` in order to run antiferromagnetic calculations? Do you need to specify two different atomic species in `StructureData`?

sphuber commented:
> @sphuber, can you help me to understand what happens for `aiida_quantumespresso` in order to run antiferromagnetic calculations? Do you need to specify two different atomic species in `StructureData`?

Yes, if for example you want to make anti-ferromagnetic iron, you will need to create a `StructureData` with two different kind of iron, `Fe1` and `Fe2`. Then you can specify in the parameters `parameters = {'ELECTRONS': {'starting_magnetization': {'Fe1': 0.5, 'Fe2': -0.5}}}`

bosonie commented:
Introduced a draft for a class (`InitMagnetizationManager`) that helps to manage the agreed format of `initial_magnetization` (Bonan suggestion on wiki).
For instance, I added a method that returns a dictionary mapping kinds to an initial magnetization. This methods also return a modified structure (unstored), where the kinds has been modified to allow complicated magnetization arrangements also for codes that require to pass magnetization for each kind.
This is just an example of useful methods we can collect on this `InitMagnetizationManager`. Fell free to suggest more (for sure some change of units will be involved)

bosonie commented:
My idea is to collect in this manager methods that developers can use inside their `get_builder` in order to process the `initial_magnetization` required by the common API.

bosonie commented:
Very sorry to bother you @sphuber and @zhubonan, but it would be great to have feedback within the day, in order to allow more time as possible to other developers for their implementations. There will be time later for further improvements, but we should make available the core structure as soon as possible.

sphuber commented on aiida_common_workflows/workflows/relax/generator.py:
```diff
@@ -68,28 +85,49 @@ def get_builder(
         previous_workchain=None,
+        is_insulator=False,
+        spin=SpinType.NONE,
+        initial_magnetization='auto',
```
I tried to find the discussion on the decision to use a string as default and to mean auto-assign, but not sure if this is ideal. Apart from it mixing types (list and string) I think it could be simply inferred from the `spin` argument. I would set the default to `None` and just accept an explicit list as alternative. If it is `None` *and* `spin != SpinType.NONE` then the builder should automatically determine it. I think it never makes sense to have a spin calculation without setting an initial magnetization.

bosonie replied:
Agreed

sphuber commented on aiida_common_workflows/workflows/relax/generator.py:
```diff
@@ -68,28 +85,49 @@ def get_builder(
         :param threshold_forces: target threshold for the forces in eV/Å.
         :param threshold_stress: target threshold for the stress in eV/Å^3.
         :param previous_workchain: a <Code>RelaxWorkChain node.
+        :param is_insulator: a bool to activate insulator options (by default metal calcs are assumed.
```
```suggestion
        :param is_insulator: set to `True` to treat the system as an insulator, default is `False`.
```

sphuber commented on aiida_common_workflows/workflows/relax/generator.py:
```diff
@@ -68,28 +85,49 @@ def get_builder(
         :param threshold_stress: target threshold for the stress in eV/Å^3.
         :param previous_workchain: a <Code>RelaxWorkChain node.
+        :param is_insulator: a bool to activate insulator options (by default metal calcs are assumed.
+        :param spin: the spin state of the calculation, instance of `RelaxType`.
```
```suggestion
        :param spin: the spin polarization type to use for the calculation, instance of `SpinType`.
```

sphuber commented on aiida_common_workflows/workflows/relax/generator.py:
```diff
@@ -68,28 +85,49 @@ def get_builder(
         :param previous_workchain: a <Code>RelaxWorkChain node.
+        :param is_insulator: a bool to activate insulator options (by default metal calcs are assumed.
+        :param spin: the spin state of the calculation, instance of `RelaxType`.
+        :param initial_magnetization: a list with the spin polarization for each site.
```
```suggestion
        :param initial_magnetization: a list with the initial spin polarization for each site. If not defined, the builder will automatically define the initial magnetization if and only if `spin != SpinType.NONE`.
```
We should also add the units here. I think you wrote elsewhere that it is the difference in electrons in up channel vs down channel? So it is a float in units of electrons? Does this also make sense for noncollinear calculations? Do we need more information there?

bosonie replied:
Ok units. We decided to deal with non-collinear cases in the future, not now. For sure we need angles.

sphuber replied:
If we will need angles in the future, we should already think how they will have to be passed in the future. I have never used noncollinearity myself so I am not sure how this is normally defined, so cannot necessarily suggest anything. Would this require a different format for `initial_magnetization`? Or is an additional argument required? Is there any interface that we can choose now that will make it easy to extend in the future?

bosonie replied:
I think the extension will simply be that the list will not accept floats or integers, but a dictionary or a list containing more info per site. The interface will remain the same and the current implementation is good. We skip your other suggestion of checking if the list contain floats or integer and we then just have to decide what to accept.

zhubonan replied:
Perhaps when a list of list is passed we can interpret it as  vectors on each site, and when a list of dict is passed the keys of the dictionary will be used, such as `angle`, `magnitude` etc.

bosonie replied:
This is something we have to agree on, but the implementation will not change, except maybe adding few checks. I will write an issue.

sphuber commented on aiida_common_workflows/workflows/relax/generator.py:
```diff
@@ -68,28 +85,49 @@ def get_builder(
+            raise ValueError('relaxation type `{}` is not supported'.format(relaxation_type))
+
+        if is_insulator not in [False, True, None]:
+            raise ValueError('The argument `is_insulator` accepts only `False`, `True` or `None`')
```
If we accept `None`, what is that supposed to mean? What is the difference with `False`? If we do accept this value, maybe that should be the default. Or maybe we should also switch this flag to an enum like `spin`, something like `electronic_type` with an `ElectronicType` enum with for now the fields `INSULATOR`, `METAL`. This will give us more flexibility in the future to extend it without risking to having to break the interface

bosonie replied:
None signal cases when the implementation can automatically decide whether to run a calculation with metal option or not. I suggested "auto" during last meeting, but in order to not mix string you suggested None. Because not all implementations can automatically decide whether to run a calculation with metal option, the default must be False, not None. I don't see the reason to have an `ElectronicType`, there aren't more options to cover I guess. Do you see some?

sphuber replied:
well there is `INSULATOR`, `METAL` and `NONE`. The `NONE` is a real third option in this context as in this means that the workflow will have to try and determine it automatically which is very different from the other two. I think having it as an explicit value makes it a lot more intuitive and safe that `None`. I may have suggested this, but only as an improvement over using a string. On top of that, could it be that in the future (some) plugins may want or need to specify for example `SEMI_METAL` or `SEMI_CONDUCTOR`? It may be a stretch, but by using a type, we allow for this flexibility in the future and don't have to break the interface, which would cause a lot of problems. Finally, since we already use enums for `RelaxType` and `SpinType`, I think having the same interface for `ElectronicType` makes sense and will give a homogeneous interface

bosonie replied:
Ok with me. Only thing is that this decision is deviating a bit from what we agreed at the last meeting. Shouldn't be a problem, but maybe let's leave few hours to allow others to express their opinion in case. @giovannipizzi @espenfl @sponce24 @chrisjsewell @ConradJohnston @eimrek @pzarabadip @yakutovicha @adegomme @broeder-j

giovannipizzi replied:
OK with me to have multiple values.
In this case, however, rather than `NONE` I would use something that really says what's going to happen, like `AUTOMATIC`? `NONE` is a bit unclear.

sponce24 replied:
I agree with @giovannipizzi I also prefer 'AUTOMATIC' than 'NONE'

sphuber replied:
fine for me to make it `AUTOMATIC`.

bosonie replied:
Thanks for the feedback. I implemented the enum ElectronicType and the logic related to it

sphuber commented on aiida_common_workflows/workflows/relax/generator.py:
```diff
@@ -68,28 +85,49 @@ def get_builder(
         :param threshold_stress: target threshold for the stress in eV/Å^3.
         :param previous_workchain: a <Code>RelaxWorkChain node.
+        :param is_insulator: a bool to activate insulator options (by default metal calcs are assumed.
+        :param spin: the spin state of the calculation, instance of `RelaxType`.
```
I think we should call this `spin_type` instead as it's a lot clearer

bosonie replied:
Ok

sphuber commented on aiida_common_workflows/workflows/relax/generator.py:
```diff
@@ -68,28 +85,49 @@ def get_builder(
+            raise ValueError('spin type `{}` is not supported'.format(spin))
+
+        if initial_magnetization != 'auto':
+            if not isinstance(initial_magnetization, list):
```
probably also want to check that the list is flat and contains only floats/integers

bosonie replied:
Ok

sphuber commented on tests/workflows/relax/test_generator.py:
```diff
@@ -47,17 +49,18 @@ def test_validation(protocol_registry):
 
     class InputsGenerator(protocol_registry, RelaxInputsGenerator):
-        """Invalid inputs generator implementation."""
+        """Invalid inputs generator implementation. No gey_builder"""
```
```suggestion
        """Invalid inputs generator implementation: no ``get_builder``"""
```

sphuber commented on tests/workflows/relax/test_generator.py:
```diff
@@ -47,17 +49,18 @@ def test_validation(protocol_registry):
 
     class InputsGenerator(protocol_registry, RelaxInputsGenerator):
-        """Invalid inputs generator implementation."""
+        """Invalid inputs generator implementation. No process class passed."""
```
```suggestion
        """Invalid inputs generator implementation: no process class passed."""
```

sphuber commented on tests/workflows/relax/test_generator.py:
```diff
@@ -69,10 +72,11 @@ def get_builder(self):
 
     class InputsGenerator(protocol_registry, RelaxInputsGenerator):
-        """Invalid inputs generator implementation."""
+        """Invalid inputs generator implementation. No _relax_types"""
```
```suggestion
        """Invalid inputs generator implementation: no ``_relax_types``"""
```

sphuber commented on tests/workflows/relax/test_generator.py:
```diff
@@ -81,10 +85,11 @@ def get_builder(self):
 
     class InputsGenerator(protocol_registry, RelaxInputsGenerator):
-        """Invalid inputs generator implementation."""
+        """Invalid inputs generator implementation. No calc_types"""
```
```suggestion
        """Invalid inputs generator implementation: no ``_calc_types``"""
```

sphuber commented on tests/workflows/relax/test_generator.py:
```diff
@@ -93,7 +98,20 @@ def get_builder(self):
 
     class InputsGenerator(protocol_registry, RelaxInputsGenerator):
-        """Invalid inputs generator implementation."""
+        """Invalid inputs generator implementation. No spin_types"""
```
```suggestion
        """Invalid inputs generator implementation: no ``_spin_types``"""
```

sphuber commented on tests/workflows/relax/test_generator.py:
```diff
@@ -93,7 +98,20 @@ def get_builder(self):
+        InputsGenerator(process_class=CommonRelaxWorkChain)
+
+    class InputsGenerator(protocol_registry, RelaxInputsGenerator):
+        """Invalid inputs generator implementation. Invalid _relax_types"""
```
```suggestion
        """Invalid inputs generator implementation: invalid ``_relax_types``"""
```

sphuber commented on tests/workflows/relax/test_generator.py:
```diff
@@ -104,6 +122,19 @@ def get_builder(self):
         InputsGenerator(process_class=CommonRelaxWorkChain)
 
+    class InputsGenerator(protocol_registry, RelaxInputsGenerator):
+        """Invalid inputs generator implementation. Invalid _relax_types"""
```
```suggestion
        """Invalid inputs generator implementation: invalid ``_spin_types``"""
```

sphuber commented on tests/workflows/relax/test_generator.py:
```diff
@@ -118,3 +149,8 @@ def test_get_calc_type_schema(inputs_generator):
+
+def test_get_spin_types(inputs_generator):
+    """Test `RelaxInputsGenerator.get_spin_types`."""
+    assert set(inputs_generator.get_spin_types()) == set([SpinType.NONE, SpinType.COLLINEAR])
```
```suggestion
    assert set(inputs_generator.get_spin_types()) == {SpinType.NONE, SpinType.COLLINEAR}
```

sphuber commented on tests/workflows/relax/test_generator.py:
```diff
@@ -118,3 +149,8 @@ def test_get_calc_type_schema(inputs_generator):
 def test_get_relaxation_types(inputs_generator):
     """Test `RelaxInputsGenerator.get_relaxation_types`."""
     assert set(inputs_generator.get_relaxation_types()) == set([RelaxType.ATOMS, RelaxType.ATOMS_CELL])
```
```suggestion
    assert set(inputs_generator.get_relaxation_types()) == {RelaxType.ATOMS, RelaxType.ATOMS_CELL}
```

sphuber commented on aiida_common_workflows/common/init_magnetization.py:
```diff
@@ -0,0 +1,116 @@
+    and contains methods to manage these quantities.
+    """
+
+    def __init__(self, structure, initial_spin_site):
```
Why `initial_spin_site` and not `initial_magnetization` just as for the builder?

bosonie replied:
Ok

zhubonan replied:
I guess it was to remind it is a list of the spin per site? `initital_magnetization` seems to be too general.

sphuber replied:
I guess we could rename it to `magnetization_per_site`, this way it is explicit and I don't think the `initial` adds anything. That is implied. It is not going to ever be the magnetization that the code is going to set halfway through

bosonie replied:
Perfect. Agreed on `magnetization_per_site`.

sphuber commented on aiida_common_workflows/common/init_magnetization.py:
```diff
@@ -0,0 +1,116 @@
+        """
+
+        if not isinstance(structure, StructureData):
+            raise RuntimeError('structure input must be a StructureData')
```
these should all be `ValueError`s

bosonie replied:
Ok

sphuber commented on aiida_common_workflows/common/init_magnetization.py:
```diff
@@ -0,0 +1,116 @@
+            raise RuntimeError('an initial magnetization for each site must be spedified')
+
+        self.structure = structure
+        self.magn_list = initial_spin_site
```
is there a reason for changing the attribute name to `magn_list`?

sphuber commented on aiida_common_workflows/common/init_magnetization.py:
```diff
@@ -0,0 +1,116 @@
+        """
+        return self.magn_list
+
+    def get_per_kind_dict_and_structure(self):
```
Two high-level comments for this function before I go into the details of the implementation:

* You are potentially returning a new structure, but it is not going through a `calcfunction` so you will lose the provenance.
* If I understand correctly, you are doing exact comparisons on the magnetization for each site to see if they are the same and so can be modeled by the same kind. Would it be sensible to include some tolerances here? i.e. if you have 0.5 and 0.5000001, I think they can safely be considered to be the same kind

On the last point, I have actually been working on these very problems and already implemented a `calcfunction` that does this for Quantum ESPRESSO, but that should be generalizable.

On that note, is this manager actually being used anywhere in the code now? Or is the idea that the plugins can use them should they want to?

bosonie replied:
>     * You are potentially returning a new structure, but it is not going through a `calcfunction` so you will lose the provenance.

The idea is that is called inside `get_builder`, therefore is not such a big deal to lose the provenance, actually everything that happens in `get_builder` is lost. Only the inputs of the workchain remain.


>     * If I understand correctly, you are doing exact comparisons on the magnetization for each site to see if they are the same and so can be modeled by the same kind. Would it be sensible to include some tolerances here? i.e. if you have 0.5 and 0.5000001, I think they can safely be considered to be the same kind

Good suggestion


> On the last point, I have actually been working on these very problems and already implemented a `calcfunction` that does this for Quantum ESPRESSO, but that should be generalizable.


Fell free to replace with your implementation

> On that note, is this manager actually being used anywhere in the code now? Or is the idea that the plugins can use them should they want to?

The idea is to make it available for developers, if they want to use it.

zhubonan replied:
I think it is intended for the code like Quantum ESPRESSO that accepts the `{'kind_name': spin}` type input. So the plugins decide if they want it so not.

Another comment - yes the creation of a new `StructureData` should be wrapped in a `calcfunction`. But I think that the creation of a new structure may not be necessary if the input structure already has the required types matching the list of magnetic moments. Perhaps this can be moved to a standalone function and the unwrapped version can be called first - if the new `StructureData` contains new kinds then we actually call the wrapped version - otherwise just return the input `StructureData`. 

Yes if QuantumESPRESSO plugin has it implemeted we can steal and generalizble it from there.

sphuber replied:
> The idea is that is called inside `get_builder`, therefore is not such a big deal to lose the provenance, actually everything that happens in `get_builder` is lost.

Yes, but the `structure` is a real exception. While all other inputs of the builder are normal non-node inputs, the structure is an actual node. If you pass it as the structure to relax, and the builder silently creates a new structure without a link, then the relaxed structure and the rest of the results of the workchain are not linked to the structure you passed in. This is definitely not what we want and not what a user would expect.

sphuber replied:
@bosonie  Since this code is not actually used or tested in this PR yet, and it might need some more discussion, would it make sense to take it out of this PR so we can get the rest merged and available for the plugin developers?

bosonie replied:
yes

sphuber reviewed (changes requested):
Thanks @bosonie : mostly minor suggestions, but some name change/logic suggestions as well. And one big point about the initialization manager. If you want to discuss in person, I have some time this afternoon if that makes it easier.

zhubonan reviewed (changes requested):
Looks fine to me!
I think adding some test for the `InitMagnetizationManager` is needed though. 
We should definitely have it wrapped in a `calcfunction` if a new structure is created as a result.

bosonie commented:
Ready @sphuber @zhubonan

sphuber commented:
@bosonie are the tests supposed to fail ? If not I will wait with the final review for now

bosonie commented:
Yes, they are supposed to fail, because I didn't implement `_spin_types` and `_electronic_types` for each code. The code is ready to be reviewed

sphuber commented:
I don't think it is a good idea to start merging failing builds though. For now its fine to add `_spin_types = ()` in all the plugin generators so at least the tests pass. It is fine that the plugins later update it to actually add those types they support. Do the same for `_electronic_types` if that is necessary for the tests. I will start to review in the meantime

sphuber commented on aiida_common_workflows/workflows/relax/abinit/generator.py:
```diff
@@ -10,7 +10,7 @@
 from aiida import engine, orm, plugins
 from aiida.common import exceptions
-from aiida_common_workflows.workflows.relax.generator import (RelaxInputsGenerator, RelaxType)
+from aiida_common_workflows.workflows.relax.generator import (RelaxInputsGenerator, RelaxType, SpinType, ElectronicType)
```
```suggestion
from ..generator import RelaxInputsGenerator, RelaxType, SpinType, ElectronicType
```

sphuber commented on aiida_common_workflows/workflows/relax/generator.py:
```diff
@@ -21,15 +21,35 @@ class RelaxType(Enum):
     """Input generator for the common structure relaxation workchains.
 
-    Subclasses should define the `_calc_types` and `_relax_types` class attributes, as well as the `get_builder` method.
+    Subclasses should define the `_calc_types`, `_spin_types` and `_relax_types` class attributes,
```
```suggestion
    Subclasses should define the `_calc_types`, `_spin_types`, `_electronic_types` and `_relax_types` class attributes,
```

sphuber commented on aiida_common_workflows/workflows/relax/generator.py:
```diff
@@ -50,9 +70,21 @@ def raise_invalid(message):
         if any([not isinstance(relax_type, RelaxType) for relax_type in self._relax_types]):
             raise_invalid('`_relax_types` are not all an instance of `RelaxType`')
 
+        if any([not isinstance(spin_type, SpinType) for spin_type in self._spin_types]):
```
```suggestion
        if any(not isinstance(spin_type, SpinType) for spin_type in self._spin_types):
```

sphuber commented on aiida_common_workflows/workflows/relax/generator.py:
```diff
@@ -50,9 +70,21 @@ def raise_invalid(message):
+        if any([not isinstance(spin_type, SpinType) for spin_type in self._spin_types]):
+            raise_invalid('`_spin_types` are not all an instance of `SpinType`')
+
+        if any([not isinstance(electronic_type, ElectronicType) for electronic_type in self._electronic_types]):
```
```suggestion
        if any(not isinstance(electronic_type, ElectronicType) for electronic_type in self._electronic_types):
```

sphuber commented on aiida_common_workflows/workflows/relax/generator.py:
```diff
@@ -50,9 +70,21 @@ def raise_invalid(message):
+        if self._electronic_types is None:
+            raise_invalid('does not define `_electronic_types`.')
+
         if any([not isinstance(relax_type, RelaxType) for relax_type in self._relax_types]):
```
```suggestion
        if any(not isinstance(relax_type, RelaxType) for relax_type in self._relax_types):
```

sphuber commented on aiida_common_workflows/workflows/relax/generator.py:
```diff
@@ -104,3 +160,11 @@ def get_calc_type_schema(self, key):
+        return list(self._spin_types.keys())
+
+    def get_electronic_types(self):
+        """Return the available spin types for this input generator."""
```
```suggestion
        """Return the available electronic types for this input generator."""
```

sphuber commented on aiida_common_workflows/workflows/relax/siesta/generator.py:
```diff
@@ -4,7 +4,7 @@
 from aiida.common import exceptions
 from aiida.orm import Group
-from aiida_common_workflows.workflows.relax.generator import RelaxInputsGenerator, RelaxType
+from aiida_common_workflows.workflows.relax.generator import RelaxInputsGenerator, RelaxType, SpinType, ElectronicType
```
```suggestion
from ..generator import RelaxInputsGenerator, RelaxType, SpinType, ElectronicType
```

sphuber reviewed (changes requested):
Thanks @bosonie , just some minor changes. And as I said in another comment I would really just add empty tuples for the plugin generators so that the tests pass. Should be simple copy pasting.

bosonie commented:
Thanks @sphuber. I added a `_electronic_types` and `_spin_type` to every code-specific generator with the values we require each implementation to support, meaning spin NONE and COLLINEAR and support for METAL and INSULATOR.
Accepted all your minor changes.

sphuber commented:
Thanks a lot @bosonie !
